apple/internal/resource_rules/apple_precompiled_resource_bundle.bzl

```diff
@@ -0,0 +1,278 @@
+# limitations under the License.
+
+"""Implementation of apple_precompiled_resource_bundle rule."""
+
+load(
+    "@bazel_skylib//lib:dicts.bzl",
+    "dicts",
+)
+load(
+    "@bazel_skylib//lib:partial.bzl",
+    "partial",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:apple_toolchains.bzl",
+    "AppleMacToolsToolchainInfo",
+    "AppleXPlatToolsToolchainInfo",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:features_support.bzl",
+    "features_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:platform_support.bzl",
+    "platform_support",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:resources.bzl",
+    "resources",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:rule_attrs.bzl",
+    "rule_attrs",
+)
+load(
+    "@build_bazel_rules_apple//apple/internal:rule_support.bzl",
+    "rule_support",
+)
+load(
+    "//apple/internal:apple_product_type.bzl",
+    "apple_product_type",
+)
+
+def _apple_precompiled_resource_bundle_impl(ctx):
+    # Owner to attach to the resources as they're being bucketed.
+    owner = str(ctx.label)
+    bucketize_args = {}
+
+    rule_descriptor = rule_support.rule_descriptor(
+        platform_type = str(ctx.fragments.apple.single_arch_platform.platform_type),
+        product_type = apple_product_type.application,
+    )
+
+    features = features_support.compute_enabled_features(
+        requested_features = ctx.features,
+        unsupported_features = ctx.disabled_features,
+    )
+
+    apple_xplat_toolchain_info = ctx.attr._xplat_toolchain[AppleXPlatToolsToolchainInfo]
+
+    platform_prerequisites = platform_support.platform_prerequisites(
+        apple_fragment = ctx.fragments.apple,
+        build_settings = apple_xplat_toolchain_info.build_settings,
+        config_vars = ctx.var,
+        cpp_fragment = ctx.fragments.cpp,
+        device_families = rule_descriptor.allowed_device_families,
+        explicit_minimum_deployment_os = None,
+        explicit_minimum_os = None,
+        features = features,
+        objc_fragment = ctx.fragments.objc,
+        platform_type_string = str(ctx.fragments.apple.single_arch_platform.platform_type),
+        uses_swift = False,
+        xcode_version_config = ctx.attr._xcode_config[apple_common.XcodeVersionConfig],
+    )
+
+    bundle_name = "{}.bundle".format(ctx.attr.bundle_name or ctx.label.name)
+    bundle_id = ctx.attr.bundle_id or None
+
+    apple_resource_infos = []
+    process_args = {
+        "actions": ctx.actions,
+        "apple_mac_toolchain_info": ctx.attr._mac_toolchain[AppleMacToolsToolchainInfo],
+        "bundle_id": bundle_id,
+        "product_type": rule_descriptor.product_type,
+        "rule_label": ctx.label,
+    }
+
+    infoplists = resources.collect(
+        attr = ctx.attr,
+        res_attrs = ["infoplists"],
+    )
+
+    if not infoplists:
+        infoplists = resources.collect(
+            attr = ctx.attr,
+            res_attrs = ["_fallback_infoplist"],
+        )
```

Comment:
I think it would be more efficient to check if `ctx.files.info.plists` is not empty, then do the first way, otherwise the second way. `resources.collect` is a little heavier.